staging/src/k8s.io/apiextensions-apiserver/pkg/controller/establish/establishing_controller.go

```diff
@@ -73,39 +73,54 @@ func (ec *EstablishingController) QueueCRD(key string, timeout time.Duration) {
 	ec.queue.AddAfter(key, timeout)
 }
 
-// Run starts the EstablishingController.
-func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+// RunWithContext starts the EstablishingController.
+func (ec *EstablishingController) RunWithContext(ctx context.Context) {
 	defer utilruntime.HandleCrash()
 	defer ec.queue.ShutDown()
 
-	klog.Info("Starting EstablishingController")
-	defer klog.Info("Shutting down EstablishingController")
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Starting EstablishingController")
+	defer logger.V(2).Info("Shutting down EstablishingController")
 
-	if !cache.WaitForCacheSync(stopCh, ec.crdSynced) {
+	if !cache.WaitForCacheSync(ctx.Done(), ec.crdSynced) {
 		return
 	}
 
-	// only start one worker thread since its a slow moving API
-	go wait.Until(ec.runWorker, time.Second, stopCh)
+	// only start one worker thread since the EstablishingController is not a bottleneck
+	go wait.UntilWithContext(ctx, ec.runWorker, time.Second)
+
+	<-ctx.Done()
+}
 
-	<-stopCh
+// Contextual logging: RunWithContext should be used instead of Run in code which supports contextual logging.
+func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
You can use `ctx := wait.ContextForChannel(stopCh)` here - no need to implement it yourself... :smile:

Reply:
Thanks for the feedback! I've made the changes and  also found and updated the caller to use the new RunWithContext method.